core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -268,13 +326,73 @@ class LogSegment(val log: FileMessageSet,
     catch {
       case e: IOException => throw kafkaStorageException("index", e)
     }
+    try timeIndex.renameTo(new File(CoreUtils.replaceSuffix(timeIndex.file.getPath, oldSuffix, newSuffix)))
+    catch {
+      case e: IOException => throw kafkaStorageException("timeindex", e)
+    }
+  }
+
+  /**
+   * Append the last time index entry to the time index when this log segment become inactive segment.
+   * This entry will be used to decide when to delete the segment.
+   */
+  def onBecomeInactiveSegment(): Unit = {
+    if (maxTimestampSoFar >= 0)
+      timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp, skipFullCheck = true)
+    else
+      timeIndex.maybeAppend(lastModified, nextOffset() - 1, skipFullCheck = true)
+  }
+
+  /**
+   * The time this segment has waited to be rolled over. If the segment does not contain any message with a timestamp,
+   * the time is based on the create time of the segment. Otherwise the time is based on the max timestamp in this
+   * segment.
+   */
+  def timeWaitedForRoll(now: Long) = now - { if (maxTimestampSoFar >= 0) maxTimestampSoFar else created}
+
+  /**
+   * Search the message offset based on timestamp.
+   * This method returns a TimestampOffset. The offset is the offset of the first message whose timestamp is
+   * greater than or equals to the target timestamp.
+   *
+   * If all the message in the segment have smaller timestamps, the offset will be last offset + 1 and the
+   * timestamp will be max timestamp in the segment.
+   *
+   * If all the messages in the segment have larger timestamps, the timestamp will be Message.NoTimestamp and
+   * the offset will be the base offset of the segment.
+   *
+   * @param timestamp The timestamp to search for.
+   * @return a TimestampOffset which points to the first message whose timestamp is larger than or equals to the
+   *         target timestamp.
+   */
+  def findOffsetByTimestamp(timestamp: Long): TimestampOffset = {
+    val timestampOffset = timeIndex.lookup(timestamp)
+    val position = index.lookup(timestampOffset.offset).position
+    log.searchForTimestamp(timestamp, position) match {
+      case Some(foundTimestampOffset) =>
+          foundTimestampOffset
+
+      case None =>
+        // That means starting offset has gone beyond the last offset in the segment.
+        // In this case we return the starting offset.
+          timestampOffset
```

Comment:
If ```log.searchForTimestamp()``` returns None, that means there is no message at or after the position we started to search with. It could happen when:
1. The log segment is empty
2. The log segment is truncated after we got the position from the offset index.

For case (1) maybe we should just return (NoTimestamp, baseOffset). For case (2) maybe we should propagate the None to the caller so the caller can retry the search. This does not completely solve the problem that we may return an offset to the user that is later truncated, but would avoid it as much as possible. Does that sound reasonable?